swap/src/bin/swap.rs

```diff
@@ -556,6 +556,11 @@ where
 {
     tracing::debug!("Requesting quote");
     let bid_quote = bid_quote.await?;
+
+    if bid_quote.max_quantity == bitcoin::Amount::ZERO {
+        bail!("The Monero seller you contacted can't trade monero at the moment")
```

Comment:
I know you wrote this from a user's perspective but in the context of this function, I'd like a different error message here. Ideally, we would create a custom error type like `struct ZeroQuote;` and return that and react to that outside of this function.

If you don't want to go that far, I'd suggest just bailing with something like:

```suggestion
        bail!("Received quote with `max_quantity` of 0");
```

And adding a `.context` to the call of `determine_btc_to_swap` here: https://github.com/comit-network/xmr-btc-swap/blob/8f3f4d5895409775506eb256f35e945da1e52de6/swap/src/bin/swap.rs#L110

This will add a "layer" to the error message that should give enough context to the user of what is going wrong. (Read about [anyhow](https://docs.rs/anyhow/latest/anyhow/index.html) if you are not familiar with it.)

Reply:
The structure of the Custom error how would you do it? with the same structure as BidQuote?

Reply:
In a nutshell, it works like this:

First, we need to make the high-level decision of the function signature:

- Do we want to use `anyhow::Result` or `std::Result`? `anyhow` provides a boxed error type which can hold anything that implements `std::error::Error`. That is convenient but reduces type-safety a bit because we no longer know exactly, which errors are returned from the function.

At the moment, we use `anyhow::Result` and I'd suggest to keep it just to keep the code-churn of this PR low. If you research about `anyhow` and Rust error handling in general, you should find a few good resources and opinions.

With `anyhow::Error` (the error type of `anyhow::Result`), it is not quite as obvious how to do type-safe error handling. What we can do is:

1. Define a new error struct: `struct ZeroQuoteReceived;` that `derive`s `thiserror::Error` or implements `std::error::Error` manually.
2. Use `bail!(ZeroQuoteReceived)` at the above place.
3. On the call-site of `determine_btc_to_swap`, we can `match` on the returned `Result` (instead of `?` it) and then use the [`anyhow::Error::downcast` API](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.downcast) to check whether the returned `anyhow::Error` "contains" a the `ZeroQuoteReceived` error type. If the `downcast` returns `Ok`, you can then handle that specific case by printing a nice message to the user. In the `Err` case, we can just `return` the error which is equivalent to what the current `?` does.

Reply:
nice!, didn't knew about downcast